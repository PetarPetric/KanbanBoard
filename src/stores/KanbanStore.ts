import { nextTick, ref } from 'vue';
import { defineStore } from 'pinia';
import type { IColumn } from "../types/Column"
import type { Toast } from "../types/Toast";
import type { ITask } from "../types/Task";
import Repository from "@/services/apiService/APIFactory";
const ColumnRepository = Repository.get("columns");
const TasksRepository = Repository.get("tasks");

export const useKanbanStore = defineStore('KanbanStore', () => {
  const listOfColumns = ref<IColumn[]>([]);
  const toast = ref<null | Toast>(null);
  const columnUpdated = ref(null);

  // Call toast component
  const setToast = (type: string, text: string) => {
    toast.value = {
      type,
      text
    };

    setTimeout(() => {
      toast.value = null;
    }, 1500);
  }

  // Fetch columns
  const fetchColumns = async () => {
    const { data } = await ColumnRepository.get()
    listOfColumns.value = data.sort(
      (a: IColumn, b: IColumn) => a.columnPosition - b.columnPosition
    )
  }

  // Add column api call 
  const addColumn = async (name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
    }
    await ColumnRepository.create(form);
    await fetchColumns();
    setToast('success', 'Successfully added column');
  }

  // Remove column
  const removeColumn = async (id?: number) => {
    await ColumnRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed column');
  }

  // rearrange columns when doing drag and drop like in jira
  const reArrangeColumns = async (id: number, columnPosition: number) => {
    const form = {
      columnPosition
    }
    await ColumnRepository.patch(form, id);
    setToast('info', 'Successfully rearranged columns');
  }

  // Edit columns
  const editColumn = async (id: number, name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
      id
    }
    await ColumnRepository.update(form, id);
    await fetchColumns();
    setToast('info', 'Edited column');
  }

  // Add task api call
  const addTask = async (columnId: number, title: string, taskPosition: number) => {
    const form: ITask = {
      title,
      description: '',
      columnId,
      taskPosition,
    }
    await TasksRepository.create(form);
    setToast('success', 'Successfully added task');
  }

  // Fetch tasks
  const fetchTasks = async (columnId: number) => {
    const { data } = await TasksRepository.get(columnId)
    return data;
  }

  const fetchSingleTask = async (id: number) => {
    const { data } = await TasksRepository.getSingle(id)
    return data;
  }

  // Move task actions
  const moveTask = async (id: number, columnId?: number, taskPosition?: number) => {
    const form = {
      columnId,
      taskPosition
    }
    const { data } = await TasksRepository.patch(form, id);
    setToast('info', 'Successfully moved task');
    return data;
  }

  // Remove task
  const removeTask = async (id: number) => {
    await TasksRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed a task')
  }

  // Edit task
  const editTask = async (form: ITask) => {
    const { data } = await TasksRepository.update(form);
    setToast('info', 'Edited Task');
    return data;
  }

  return {
    fetchColumns,
    addColumn,
    removeColumn,
    editColumn,
    reArrangeColumns,
    addTask,
    fetchTasks,
    fetchSingleTask,
    removeTask,
    moveTask,
    editTask,
    setToast,
    toast,
    listOfColumns,
    columnUpdated
  }
});