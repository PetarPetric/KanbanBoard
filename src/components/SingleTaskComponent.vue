<script setup lang="ts">
import { useKanbanStore } from "@/stores/KanbanStore";
import type { ITask } from "@/types/Task";
import { computed } from "@vue/reactivity";
import { BIconTrash } from "bootstrap-icons-vue";
import { useRouter } from "vue-router";
import { useDragAndDrop } from "@/composables/dnd";
// Emit events
const emit = defineEmits(["removedTask"]);
// Router
const $router = useRouter();
// Store
const store = useKanbanStore();
// Props
const props = defineProps<{
  singleTask: ITask;
  taskIndex: number;
  taskColumn: ITask[];
}>();

const {
  startDrag,
  onDragOver,
  dropElement,
  dragStartItem,
  goingOverItem,
  draggingItemType,
} = useDragAndDrop();

// Get title of column in task
const columnTitle = computed(() => {
  return store.listOfColumns.find(
    (item) => item.id == props.singleTask.columnId
  );
});

// Go To task details
const openTaskDetails = () => {
  $router.push({ path: `/task/${props.singleTask.id}` });
};

// Remov task function
const removeTask = async () => {
  await store.removeTask(props.singleTask.id as number);
  emit("removedTask");
};
</script>

<template>
  <div class="placeholder">
    <div
      draggable="true"
      @dragstart.stop="startDrag(props.singleTask, $event)"
      @dragover.stop="onDragOver(props.singleTask, props.taskIndex, $event)"
      @dragend.stop="
        dropElement(props.singleTask, props.taskIndex, props.taskColumn)
      "
      @click="openTaskDetails"
      data-test="single-task"
      class="single-task"
      :class="[
        {
          'drag-item':
            props.singleTask.id === dragStartItem?.id &&
            goingOverItem &&
            draggingItemType.task,
        },
      ]"
    >
      <div data-test="single-task-header" class="single-task__header">
        <span class="text-secondary">{{ props.singleTask.title }}</span>
        <button
          data-test="btn-remove"
          @click.stop="removeTask"
          class="btn btn--remove"
        >
          <BIconTrash />
        </button>
      </div>
      <div data-test="single-task-body" class="single-task__body">
        <span class="text-secondary">{{ columnTitle?.title }} column</span>
      </div>
    </div>
  </div>
</template>

<style lang="scss" scoped>
@import "@/styles/task.scss";
.placeholder {
  border-radius: 3px;
  margin: 1rem 0;
  height: 80px;
  position: relative;
  background-color: transparent;

  .single-task {
    position: absolute;
  }
}
.down {
  transform: translateY(80px);
}

.up {
  transform: translateY(-80px);
}
</style>
