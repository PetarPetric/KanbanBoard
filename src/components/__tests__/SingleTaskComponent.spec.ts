import { mount, VueWrapper, flushPromises } from "@vue/test-utils";
import { createTestingPinia } from '@pinia/testing';
import { useKanbanStore } from "@/stores/KanbanStore";
import SingleTaskComponent from "@/components/SingleTaskComponent.vue";
import { createRouterMock, injectRouterMock } from 'vue-router-mock';

describe("single task component", () => {
  let wrapper: VueWrapper<any>;
  let store: any;
  const router = createRouterMock()
  beforeEach(() => {
    injectRouterMock(router);
    wrapper = mount(SingleTaskComponent as any, {
      shallow: true,
      props: {
        singleTask: {
          id: 5,
          title: 'Just another task',
          description: '',
          columnId: 5,
          taskPosition: 0,
        },
      },
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn,
          initialState: {
            KanbanStore: {
              listOfColumns: [
                {
                  title: "Another Task title",
                  boardId: 0,
                  id: 5
                },
                {
                  title: "Random title",
                  boardId: 0,
                  id: 2
                },
                {
                  title: "New Title",
                  boardId: 0,
                  id: 1
                },
              ],
            }
          }
        })]
      }
    });

    store = useKanbanStore();
  });

  afterEach(() => {
    wrapper.unmount()
  })

  it("should render correctly", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call delete method', async () => {
    const btnRemove = wrapper.find('[data-test="btn-remove"]');
    await btnRemove.trigger('click')
    expect(store.removeTask).toHaveBeenCalledTimes(1)
    expect(store.removeTask).toHaveBeenLastCalledWith(wrapper.vm.props.singleTask.id)
  })

  it('should redirect to modal', async () => {
    await wrapper.find('[data-test="single-task"]').trigger('click');
    expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
    expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith({ path: `/task/${wrapper.vm.props.singleTask.id}` });
    await wrapper.vm.$router.push(`/task/${wrapper.vm.props.singleTask.id}`);
    expect(wrapper.vm.$route.path).toBe(`/task/${wrapper.vm.props.singleTask.id}`)
    expect(wrapper.find('.modal__wrapper')).toBeTruthy()
  })
});
