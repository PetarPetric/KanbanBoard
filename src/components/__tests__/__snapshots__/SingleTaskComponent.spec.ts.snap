// Vitest Snapshot v1

exports[`single task component > should render correctly 1`] = `
"<div class=\\"placeholder\\" data-v-f2269eac=\\"\\">
  <div draggable=\\"true\\" data-test=\\"single-task\\" class=\\"single-task\\" data-v-f2269eac=\\"\\">
    <div data-test=\\"single-task-header\\" class=\\"single-task__header\\" data-v-f2269eac=\\"\\"><span class=\\"text-secondary\\" data-v-f2269eac=\\"\\">Just another task</span><button data-test=\\"btn-remove\\" class=\\"btn btn--remove\\" data-v-f2269eac=\\"\\">
        <b-icon-trash-stub data-v-f2269eac=\\"\\"></b-icon-trash-stub>
      </button></div>
    <div data-test=\\"single-task-body\\" class=\\"single-task__body\\" data-v-f2269eac=\\"\\"><span class=\\"text-secondary\\" data-v-f2269eac=\\"\\">Another Task title column</span></div>
  </div>
</div>"
`;
